src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -341,6 +381,706 @@ CreateInsertSelectRouterPlan(Query *originalQuery,
+ *
+ * Basically, the function iterates over the secondClass' member list and checks whether
+ * it already exists in the firstClass' member list. If not, the member is added to the
+ * firstClass.
+ */
+static void
+ListConcatUniqueAttributeClassMemberLists(AttributeEquivalenceClass **firstClass,
+										  AttributeEquivalenceClass *secondClass)
+{
+	ListCell *equivalenceClassMemberCell = NULL;
+	List *equivalenceMemberList = secondClass->equivalentAttributes;
+
+	foreach(equivalenceClassMemberCell, equivalenceMemberList)
+	{
+		AttributeEquivalenceClassMember *newEqMember = lfirst(equivalenceClassMemberCell);
+
+		if (AttributeClassContainsAttributeClassMember(newEqMember, *firstClass))
+		{
+			continue;
+		}
+
+		(*firstClass)->equivalentAttributes = lappend((*firstClass)->equivalentAttributes,
+													  newEqMember);
+	}
+}
+
+
+/*
+ * GenerateAttributeEquivalencesForJoinRestrictions gets a join restriction
+ * context and returns a list of AttrributeEquivalenceClass.
+ *
+ * The algorithm followed can be summarized as below:
+ *
+ * - Per join restriction
+ *     - Per RestrictInfo of the join restriction
+ *     - Check whether the join restriction is in the form of (Var1 = Var2)
+ *         - Create an AttributeEquivalenceClass
+ *         - Add both Var1 and Var2 to the AttributeEquivalenceClass
+ *
+ * Note that this function does not deal with whether the member of restriction
+ * are partition key or not. That's handled later in the planning.
+ */
+static List *
+GenerateAttributeEquivalencesForJoinRestrictions(JoinRestrictionContext *
+												 joinRestrictionContext)
+{
+	List *attributeEquivalenceList = NIL;
+	ListCell *joinRestrictionCell = NULL;
+
+	foreach(joinRestrictionCell, joinRestrictionContext->joinRestrictionList)
+	{
+		JoinRestriction *joinRestriction = lfirst(joinRestrictionCell);
+		ListCell *restrictionInfoList = NULL;
+
+		foreach(restrictionInfoList, joinRestriction->joinRestrictInfoList)
+		{
+			RestrictInfo *rinfo = (RestrictInfo *) lfirst(restrictionInfoList);
+			OpExpr *restrictionOpExpr = NULL;
+			Node *leftNode = NULL;
+			Node *rightNode = NULL;
+			Expr *strippedLeftExpr = NULL;
+			Expr *strippedRightExpr = NULL;
+			Var *leftVar = NULL;
+			Var *rightVar = NULL;
+			Expr *restrictionClause = rinfo->clause;
+			AttributeEquivalenceClass *attributeEquivalance = NULL;
+
+			if (!IsA(restrictionClause, OpExpr))
+			{
+				continue;
+			}
+
+			restrictionOpExpr = (OpExpr *) restrictionClause;
+			if (list_length(restrictionOpExpr->args) != 2)
+			{
+				continue;
+			}
+			if (!OperatorImplementsEquality(restrictionOpExpr->opno))
+			{
+				continue;
+			}
+
+			leftNode = linitial(restrictionOpExpr->args);
+			rightNode = lsecond(restrictionOpExpr->args);
+
+			/* we also don't want implicit coercions */
+			strippedLeftExpr = (Expr *) strip_implicit_coercions((Node *) leftNode);
+			strippedRightExpr = (Expr *) strip_implicit_coercions((Node *) rightNode);
```

Comment:
Yes, when I create two tables that are partitioned on varchar columns, I can see that we get the following here:
```
{RELABELTYPE :arg {VAR :varno 3 :varattno 1 :vartype 1043 :vartypmod -1 :varcollid 100 :varlevelsup 0 :varnoold 3 :varoattno 1 :location 49} :resulttype 25 :resulttypmod -1 :resultcollid 100 :relabelformat 2 :location -1}
```